src/engine/strat_engine/backstore/metadata.rs

```diff
@@ -36,7 +36,54 @@ pub struct BDA {
     regions: mda::MDARegions,
 }
 
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+enum MetadataLocation {
+    Both,
+    First,
+    Second,
+}
+
 impl BDA {
+    /// Read the BDA from the device and returning a byte array of the entire area.
+    fn read<F>(f: &mut F) -> io::Result<[u8; _BDA_STATIC_HDR_SIZE]>
+    where
+        F: Read + Seek,
+    {
+        #![allow(unused_io_amount)]
+        f.seek(SeekFrom::Start(0))?;
+        let mut buf = [0u8; _BDA_STATIC_HDR_SIZE];
+
+        // FIXME: See https://github.com/stratis-storage/stratisd/pull/476
+        f.read(&mut buf)?;
+        Ok(buf)
+    }
+
+    /// Write the BDA, allowing the caller to specify which part(s) to update.
+    fn write<F>(f: &mut F, bda_buf: &[u8], which: MetadataLocation) -> io::Result<()>
+    where
+        F: Seek + SyncAll,
```

Comment:
I understand that fixes to ```BDA::read``` are being deferred until a subsequent PR. But I think what regions are zeroed should not be dependent on what location argument is passed. Here's one approach which consistently zeroes nothing and which is very simple:

```
    fn write<F>(f: &mut F, bda_buf: &[u8], which: MetadataLocation) -> Result<()>
    where
        F: Seek + SyncAll,
    {
        if which == MetadataLocation::Both || which == MetadataLocation::First {
            f.seek(SeekFrom::Start(SECTOR_SIZE as u64))?;
            f.write_all(bda_buf)?;
            f.sync_all()?;
        }

        if which == MetadataLocation::Both || which == MetadataLocation::Second {
            f.seek(SeekFrom::Start(9 * SECTOR_SIZE as u64))?;
            f.write_all(bda_buf)?;
            f.sync_all()?;
        }
        Ok(())
    }
```

and here's another approach which consistently zeroes everything and is not too complex:

```
    fn write<F>(f: &mut F, bda_buf: &[u8], which: MetadataLocation) -> Result<()>
    where
        F: Seek + SyncAll,
    {
        let zeroed = [0u8; _BDA_STATIC_HDR_SIZE];

        f.seek(SeekFrom::Start(0))?;
        f.write_all(&zeroed[..SECTOR_SIZE])?; // Zero 1 unused sector
        if which == MetadataLocation::Both || which == MetadataLocation::First {
            f.write_all(bda_buf)?;
        } else {
            f.seek(SeekFrom::Start(SECTOR_SIZE as u64 * 2))?;
        }
        f.sync_all()?;

        f.write_all(&zeroed[..SECTOR_SIZE * 7])?; // Zero 7 unused sectors
        if which == MetadataLocation::Both || which == MetadataLocation::Second {
            f.write_all(bda_buf)?;
        } else {
            f.seek(SeekFrom::Start(SECTOR_SIZE as u64 * 10))?;
        }
        f.write_all(&zeroed[..SECTOR_SIZE * 6])?; // Zero 6 unused sectors
        f.sync_all()?;
        Ok(())
    }
```

Both approaches manage to pass all tests in the metadata module, FWIW.

Reply:
@mulkieran Please elaborate on why you believe we should re-write all the zero parts, what value add does that provide?

Reply:
I don't take any stand on whether the empty parts of the BDA should be zeroed or not. I only think that what is zeroed should not be dependent on the value of the ```which``` argument. I presented two alternatives, one which zeroes nothing and one which zeroes everything.

Reply:
When we have gotten into this situation we have one copy that is good, one that is bad.

IMHO writing to both regions does not improve the odds of having both copies, but decreases it as you are re-writing something known to be good.  You never **know** if a write is actually good unless you re-read it, so if it's known to be good, don't muck with it!

We have two copies, we want to have the ability to write to both or one or the other.  Part of each copy has reserved sectors and @agrover wants reserved sectors zeroed.  I don't know why the function needs to write all or none zero parts instead of just writing just the parts the caller specified. It's bad when function calls have unintended side effects.

Reply:
I think some of the stuff in my head didn't make it into the doc! (sorry.)

So what I was thinking was that the StaticHeader was divided into 2 4K chunks. We're only using one sector from each, to prevent partial writes on 512byte blocksize disks. (And the copies aren't right next to each other, because we want them to be separate writes on 4k disks.)

If someone overwrote the first 4K with garbage (but the second one is ok), I'd like to restore the first signature block, but also re-zero-out the rest of the 4K too. Although this is a little sketchy because otherwise we don't look at the other 7 sectors at all, i.e. verify they're zero, I don't think. So maybe now that I think about it we shouldn't zero. Aw hell. Whatever you wanna do, Tony.

Reply:
What regions are known to be good when ```write``` is invoked?

Reply:
I've added a third option, which is also perfectly fine w/ me, as it is consistent w/ itself. It also seems to be what @agrover intended, according to discussion in standup.

Reply:
@mulkieran @agrover The third option looks good to me :+1:   I incorporate it with the other suggested changes.

Reply:
I've made an edit to my third choice. 8K was floating around in the comments where 8 _sectors_ was really meant.